Edit remote monitoring and visualization how-to (#1058)
AFLplusplus/AFLplusplus: docs/rpc_statsd.md

# Remote monitoring and metrics visualization

> With StatsD, Prometheus, and Grafana you can set up a tool chain for remote monitoring and visualization of AFL++ metrics. All tools are free and open source.

Reviewer: AFL++'s feature is to send out StatsD messages, using it together with Prometheus and Grafana is just one possible way to visualize things. Maybe that's important here?

Author: ```suggestion
AFL++ can send out metrics as StatsD messages. For remote monitoring and visualization of the metrics, you can set up a tool chain. For example, with Prometheus and Grafana. All tools are free and open source.
```

Does this work?

Reviewer: Sounds good :)

This enables you to create nice and readable dashboards containing all the information you need on your fuzzer instances. There is no need to write your own statistics parsing system, deploy and maintain it to all your instances, and sync with your graph rendering system.

> Compared to the default integrated UI of AFL++, this can help you to visualize trends and the fuzzing state over time. You might be able to see when the fuzzing process has reached a state of no progress and visualize what are the "best strategies" for your targets (according to your own criteria). You can do so without logging into each instance manually.

Reviewer 2: ```suggestion
Compared to the default integrated UI of AFL++, this can help you to visualize trends and the fuzzing state over time. You might be able to see when the fuzzing process has reached a state of no progress and visualize what are the "best strategies" for your targets (according to your own criteria). You can do so without logging into each instance individually.
```

maybe ?

![example visualization with Grafana](resources/statsd-grafana.png)

This is an example visualization with Grafana. The dashboard can be imported with [this JSON template](resources/grafana-afl++.json).

> ## Afl++ metrics and StatsD

Reviewer 2: ```suggestion
## AFL++ metrics and StatsD
```

    The available values are the following:
    -  `dogstatsd`
    -  `influxdb`
    -  `librato`
    -  `signalfx`
    
    For more information on environment variables, see [docs/env_variables.md](docs/env_variables.md).

    Note: When using multiple fuzzer instances with StatsD it is *strongly* recommended to set up `AFL_STATSD_TAGS_FLAVOR` to match your StatsD server. This will allow you to see individual fuzzer performance, detect bad ones, and see the progress of each strategy.

3. Optional: To set the host and port of your StatsD daemon, set `AFL_STATSD_HOST` and `AFL_STATSD_PORT`. The default values are `localhost` and `8125`.

## Installing and setting up StatsD, Prometheus, and Grafana

The easiest way to install and set up the infrastructure is with Docker and Docker Compose.

Depending on your fuzzing setup and infrastructure, you may not want to run these applications on your fuzzer instances. This setup may be modified before use in a production environment; for example, adding passwords, creating volumes for storage, tweaking the metrics gathering to get host metrics (CPU, RAM, and so on).

> For all your fuzzers, only one instance of StatsD, Prometheus, and Grafana is required.

Reviewer: StatsD messages are sent out by AFL++, it's not a tool you could have an instance of

Reviewer 2: You maybe need a metric collector (agent, remote backend...) to run (either locally or remotely).
In this case, it's the "prometheus exporter" (image: `prom/statsd-exporter` in the `docker-compose.yml`) but yea that's not "StatsD" per se.

Author: Suggestion to fix the sentence:
```suggestion
For all your fuzzing instances, only one instance of Prometheus and Grafana is required.
```

Do I need to add anything about the exporter?

Reviewer 3: @Edznux can you clarify about llzmb's questions about the exporter? thanks :)

Reviewer 3: he cant login to github at the moment, he wrote to me: "basically I think it's fine to add just something along the line of "statsd_exporter is useful when dealing with Prometheus, as it converts the statsd metrics to prometheus. If you are using a provider that support statsd directly, you can skip this part If you are going to use the provided Prometheus/graphana stack, you'll need the statsd_exporter"  "

Author: Thanks for the input!

Step 2 of the tutorial contains `image: prom/statsd-exporter` in the `docker-compose.yml`. So, we only describe one way to do things. My suggestion:

```suggestion
For all your fuzzing instances, only one instance of Prometheus and Grafana is required. The [statsd exporter](https://registry.hub.docker.com/r/prom/statsd-exporter) converts the StatsD metrics to Prometheus. If you are using a provider that supports StatsD directly, you can skip this part of the setup."
```

You can create and move the infrastructure files into a directory of your choice. The directory will store all the required configuration files.

> To install and set up StatsD, Prometheus, and Grafana:

Reviewer 2: Same here.

Author: ```suggestion
To install and set up Prometheus and Grafana:
```

Maybe?

Reviewer 3: yes
